test/HashingTests.hs

```diff
@@ -0,0 +1,90 @@
+{-# LANGUAGE CApiFFI #-}
+{-# LANGUAGE TypeApplications #-}
+
+module HashingTests where
+
+import Cryptography.Sodium.Bindings.Hashing
+import Foreign (Ptr, mallocBytes, peekArray0)
+import qualified Foreign as C
+import Foreign.C (CInt (..), CSize (..), CUChar)
+import qualified Foreign.C.String as C
+import Test.Tasty
+import Test.Tasty.HUnit
+
+spec :: TestTree
+spec =
+  testGroup
+    "Hashing tests"
+    [ testCase "Test hashing a single message with a key" testHashingWithAKey,
+      testCase "Test hashing a single message with no key" testHashingWithNoKey
+    ]
+
+testHashingWithAKey :: Assertion
+testHashingWithAKey = do
+  outPtr1 <- mallocBytes (fromIntegral cryptoGenericHashBytes)
+  let outParameter1 = C.castPtr outPtr1 :: Ptr CUChar
+  let outLen1 = cryptoGenericHashBytes
+
+  outPtr2 <- mallocBytes (fromIntegral cryptoGenericHashBytes)
+  let outParameter2 = C.castPtr outPtr2 :: Ptr CUChar
+  let outLen2 = cryptoGenericHashBytes
+
+  let messageString = "Hi Bob, I'm in your system now! ;)"
+  (cString, stringSize) <- C.newCStringLen messageString
+  let inLen = fromIntegral stringSize
+  let inParameter = C.castPtr cString :: Ptr CUChar
+
+  keyPtr <- mallocBytes (fromIntegral cryptoGenericHashKeyBytes)
+  let keyParameter = C.castPtr keyPtr :: Ptr CUChar
+  let keyLen = cryptoGenericHashKeyBytes
+  cryptoGenericHashKeyGen keyParameter
+
+  cryptoGenericHash
+    outParameter1
+    outLen1
+    inParameter
+    inLen
+    keyParameter
+    keyLen
+
+  cryptoGenericHash
+    outParameter2
+    outLen2
+    inParameter
+    inLen
+    keyParameter
+    keyLen
+
+  result <- memcmp outParameter1 outParameter2 cryptoGenericHashBytes
+
+  assertBool "The result is expected" (result == 0)
+
+testHashingWithNoKey :: Assertion
+testHashingWithNoKey = do
+  outPtr <- mallocBytes (fromIntegral cryptoGenericHashBytes)
+  let outParameter = C.castPtr outPtr :: Ptr CUChar
+  let outLen = cryptoGenericHashBytes
+
+  let messageString = "Hi Bob, I'm in your system now! ;)"
+  (cString, stringSize) <- C.newCStringLen messageString
+  let inLen = fromIntegral stringSize
+  let inParameter = C.castPtr cString :: Ptr CUChar
+
+  keyPtr <- mallocBytes (fromIntegral cryptoGenericHashKeyBytes)
+  let keyParameter = C.castPtr keyPtr :: Ptr CUChar
+  let keyLen = cryptoGenericHashKeyBytes
+
+  cryptoGenericHash
+    outParameter
+    outLen
+    inParameter
+    inLen
+    keyParameter
+    keyLen
+
+  dataArray <- peekArray0 0x0 outParameter
+  let result = toEnum @Char . fromEnum <$> dataArray
+  assertEqual "The result is expected" result "X\181\149\168\&9\179\181i\191\146*\223\222U9\190\162\229l;\251\230\b\162k\162/\248!\170H\ACK"
```

Comment:
Instead of round-tripping through `Char`, I think you can get a much more meaningful result by making an array to compare with.